### PR TITLE
Render layer name collector

### PR DESCRIPTION
**Pull request by antirotor**
## Changelog Description
This PR adds collector for render layer names - so they can be used later on in anatomy templates (either for publishing or delivery).

> [!WARNING]
Due to the dynamic nature of product templates it is difficult to determine good `group_name` from that. It should be basically `product_name` without **aov** data in it, but since the product name template might look like `{product[type]}<{Renderlayer}><_{Aov}>`, it is difficult to get it just from the resolved string.

## Testing notes:

1. With mentioned ayon-core PR enable new behavior in the settings
2. Set render product name template correctly
3. Render something (locally and on farm)

### CODE REVIEW

**Review by moonyuet — COMMENTED** [review]
Did a simple test with both local and farm rendering mantra. It is rendered successfully.
image
I will test it with different template later to double confirm

File: client/ayon_houdini/plugins/publish/collect_local_render_instances.py [hunk 1]
```diff
@@ -62,23 +65,23 @@ def process(self, instance):
...
-        )
-
-        for aov_name, aov_filepaths in expectedFiles.items():
-            product_name = product_group
 
+        for aov_name, aov_filepaths in expected_files.items():
+            dynamic_data = {}
             if aov_name:
-                product_name = "{}_{}".format(product_name, aov_name)
+                dynamic_data["aov"] = aov_name
```
Comment (MustafaJafar) [inline client/ayon_houdini/plugins/publish/collect_local_render_instances.py:75]: It misses render layer dynamic data. 

```suggestion
            dynamic_data = {
                "renderlayer": instance.data["renderlayer"]
            }
            if aov_name:
                dynamic_data["aov"] = aov_name
```

File: client/ayon_houdini/plugins/publish/collect_local_render_instances.py [hunk 2]
```diff
@@ -62,23 +65,23 @@ def process(self, instance):
...
+
+            product_name, product_group = get_product_name_and_group_from_template(
+                project_name=context.data["projectName"],
+                task_entity=context.data["taskEntity"],
+                host_name=context.data["hostName"],
+                product_type=product_type,
+                variant=instance.data["productName"],
+                dynamic_data = dynamic_data,
```
Comment (moonyuet) [inline client/ayon_houdini/plugins/publish/collect_local_render_instances.py:83]:
```suggestion
                dynamic_data=dynamic_data,
```

**Comment by MustafaJafar** [comment]
I need to say, that adding a wrong/bad value for templates name profiles can lead to this error.
image

and, they may be clueless until they open the details page.

image

So, could we do something to help about it ? 
maybe `CollectLocalRenderInstances`  message about creating instances with the same name.

**Comment by moonyuet** [comment]
> I need to say, that adding a wrong/bad value for templates name profiles can lead to this error.

This product name template is not applicable for the houdini host only, guess we should check if there is wrong template name generally.

> So, could we do something to help about it ? maybe `CollectLocalRenderInstances` message about creating instances with the same name.

From your screenshot, it seems there is something wrong when creating render instances with specific name which these variants name doesn't get check which leads to the duplicates.  It is supposed that it is always checking(it works at least when we create rop instance with the same names.

File: client/ayon_houdini/plugins/publish/collect_local_render_instances.py (hunk 1 above)
Comment (antirotor) [reply client/ayon_houdini/plugins/publish/collect_local_render_instances.py:75]: I've added *renderlayer* in optional way so it is not broken whenever this plugin changes families (and the render layer collector doesn't). Dynamic data shouldn't be mandatory anyway.

**Comment by antirotor** [comment]
I've added some help in https://github.com/ynput/ayon-core/pull/841/commits/f752940c1175cfd2bd9665264e218ed0732622c1

With product name templates (not just connected to these PRs), they need to be configured really carefuly. Any mistake can break a lot of things so it shouldn't be something artists themselves hit that often.

**Comment by MustafaJafar** [comment]
Firstly, I must have a template profile for render in Houdini, otherwise the publisher will complain about having multiple instances with the same name.
image

---

Farm doesn't use the correct naming but it works locally. 
image

Local Log: 
image

From Farm job json file: 
```
{
    .
    .
    "instances": [
        {
            "aov": "beauty",
            "productGroup": "renderCfxMantra_ropTest",
            "productName": "renderCfxMantra_ropTest_beauty",
            "productType": "render",
            "renderlayer": "mantra_ropTest",
            .
            .
        },
        {
            "aov": "N",
            "productGroup": "renderCfxMantra_ropTest",
            "productName": "renderCfxMantra_ropTest_N",
            "productType": "render",
            "renderlayer": "mantra_ropTest",
            .
            .
        },
        {
            "aov": "P",
            "productGroup": "renderCfxMantra_ropTest",
            "productName": "renderCfxMantra_ropTest_P",
            "productType": "render",
            "renderlayer": "mantra_ropTest",
            .
            .
        }
    ],
    .
    .
}
```

**Review by MustafaJafar — COMMENTED** [review]
Firstly, I must have a template profile for render in Houdini, otherwise the publisher will complain about having multiple instances with the same name.
image

---

Farm doesn't use the correct naming but it works locally. 
image

Local Log: 
image

From Farm job json file: 
```json
{
    .
    .
    "instances": [
        {
            "aov": "beauty",
            "productGroup": "renderCfxMantra_ropTest",
            "productName": "renderCfxMantra_ropTest_beauty",
            "productType": "render",
            "renderlayer": "mantra_ropTest",
            .
            .
        },
        {
            "aov": "N",
            "productGroup": "renderCfxMantra_ropTest",
            "productName": "renderCfxMantra_ropTest_N",
            "productType": "render",
            "renderlayer": "mantra_ropTest",
            .
            .
        },
        {
            "aov": "P",
            "productGroup": "renderCfxMantra_ropTest",
            "productName": "renderCfxMantra_ropTest_P",
            "productType": "render",
            "renderlayer": "mantra_ropTest",
            .
            .
        }
    ],
    .
    .
}
```

**Review by MustafaJafar — APPROVED** [review]
Hey, this works fine on my side. 
My local render and farm render are following my render product template `{product[type]}<_{Renderlayer}><_{Aov}>` 

For information this toggle works only fon farm.
- `ayon+settings://core/tools/creator/use_legacy_product_names_for_renders`
image

**Comment by antirotor** [comment]
> For information this toggle works only for farm. Should we port its [logic](https://github.com/ynput/ayon-core/blob/d00fa72892f24bf50527838761016f8a28379c98/client/ayon_core/pipeline/farm/pyblish_functions.py#L705-L730) to `CollectLocalRenderInstances`?

ah, we definitely should...

**Review by MustafaJafar — APPROVED** [review]
This now is working in local and farm. 
Both respect `use_legacy_product_names_for_renders` logic.

**Comment by antirotor** [comment]
Tested and it works indeed, thanks.

**Pull request merged**